tests/rules/property.js

```diff
@@ -516,5 +540,45 @@ eslintTester.run("property", rule, {
             ],
             parser: PATH_TO_BABEL_ESLINT,
         },
+        {
+            code: "copy = '<b>safe</b>'; copy = evil; y.innerHTML = copy;",
+            errors: [
+                {
+                    message: "Unsafe assignment to innerHTML",
+                    type: "AssignmentExpression"
+                }
+            ],
+            parserOptions: { ecmaVersion: 6 }
+        },
+        {
+            code: "var copies = '<b>safe</b>'; copies = suddenlyUnsafe; y.innerHTML = copies;",
+            errors: [
+                {
+                    message: "Unsafe assignment to innerHTML",
+                    type: "AssignmentExpression"
+                }
+            ],
+            parserOptions: { ecmaVersion: 6 }
+        },
+        {
+            code: "var copies = '<b>safe</b>'; var copies = suddenlyUnsafe; y.innerHTML = copies;",
+            errors: [
+                {
+                    message: "Unsafe assignment to innerHTML",
+                    type: "AssignmentExpression"
+                }
+            ],
+            parserOptions: { ecmaVersion: 6 }
+        },
+        {
+            code: "var copies = '<b>safe</b>'; if (monday) { copies = badness }; y.innerHTML = copies;",
+            errors: [
+                {
+                    message: "Unsafe assignment to innerHTML",
+                    type: "AssignmentExpression"
+                }
+            ],
+            parserOptions: { ecmaVersion: 6 }
+        },
```

Comment:
I did also spent some time thinking of some other unsafe expression I wanted to double-check on this PR, they were all marked as unsafe, which is good, but it may be worth to still include them explicitly in the test cases to make sure that will stay the case (e.g. if eslint internal API does suddenly change behavior).

The following two cases are the ones that I could think of right now:

- double-checking that if the variable is captured and reassigned to an unsafe value from inside a closure then it is still mark it as disallowed:

```js
        {
            code: `var copies = "<b>safe</b>";
              (() => {
                  copies = badness;
              })();
              y.innerHTML = copies;
            `,
            errors: [
                {
                    message: /Unsafe assignment to innerHTML \(variable 'copies' reassigned with unsafe value at \d+:\d+\)/,
                    type: "AssignmentExpression"
                }
            ],
            parserOptions: { ecmaVersion: 6 }
        },
```

- double-checking that the plugin does still consider unsafe if an apparently safe value is retrieved by an object property which may have been passed as a reference to some other code and changed it without being detectable from a static analysis perspective (this is already marked as disallowed because of the `MemberExpression` being handled by the last branch of the `if...else` blocks in `allowedExpression`, which does `return false` for everything not matching any of the previous conditions):

```js
        {
            code: `var obj = { prop: "<b>safe</b>" };
              doSomething(obj);
              var copies = obj.prop;
              y.innerHTML = copies;
            `,
            errors: [
                {
                    message: /Unsafe assignment to innerHTML \(variable 'copies' initialized with unsafe value at \d+:\d+\)/,
                    type: "AssignmentExpression"
                }
            ],
            parserOptions: { ecmaVersion: 6 }
        },
```

Reply:
I don't think these cases are really exploring different kinds of code flow in our linter, but I agree it makes sense to hold eslint's internal API accountable to not break these assumptions for edge cases. I'd rather thave things tested too much than too little. Thanks! :)